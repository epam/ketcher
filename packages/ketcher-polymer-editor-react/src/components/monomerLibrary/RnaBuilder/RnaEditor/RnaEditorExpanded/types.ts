import { ChangeEvent } from 'react';

export interface IRnaEditorExpandedProps {
  name: string;
  onCancel: () => void;
  onChangeName: (event: ChangeEvent<HTMLInputElement>) => void;
}
