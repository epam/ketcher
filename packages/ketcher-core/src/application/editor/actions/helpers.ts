import { Atom, Bond, Neighbor, StereoLabel, Struct } from 'domain/entities';
import { StereoValidator } from 'domain/helpers';

export function getStereoAtomsMap(
  struct: Struct,
  bonds: Array<Bond>,
  bond?: Bond,
) {
  const stereoAtomsMap = new Map();
  const correctAtomIds: Array<number> = [];

  bonds.forEach((bond: Bond | undefined) => {
    if (bond) {
      const beginNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.begin,
      );
      const endNeighs: Array<Neighbor> | undefined = struct.atomGetNeighbors(
        bond.end,
      );

      if (
        StereoValidator.isCorrectStereoCenter(
          bond,
          beginNeighs,
          endNeighs,
          struct,
        )
      ) {
        const stereoLabel = struct.atoms.get(bond.begin)?.stereoLabel;
        if (
          stereoLabel == null ||
          stereoAtomsMap.get(bond.begin)?.stereoLabel == null
        ) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: getStereoParity(bond.stereo),
            stereoLabel: stereoLabel || `${StereoLabel.Abs}`,
          });
        }
        correctAtomIds.push(bond.begin);
      } else {
        if (!correctAtomIds.includes(bond.begin)) {
          stereoAtomsMap.set(bond.begin, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
        if (!correctAtomIds.includes(bond.end)) {
          stereoAtomsMap.set(bond.end, {
            stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
            stereoLabel: null,
          });
        }
      }
    }
  });

  // in case the stereo band is flipped, changed or removed
  // TODO the duplication of the code below should be fixed, mayby by function
  if (bond) {
    if (!correctAtomIds.includes(bond.begin)) {
      stereoAtomsMap.set(bond.begin, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
    if (!correctAtomIds.includes(bond.end)) {
      stereoAtomsMap.set(bond.end, {
        stereoParity: Atom.PATTERN.STEREO_PARITY.NONE,
        stereoLabel: null,
      });
    }
  }

  return stereoAtomsMap;
}

function getStereoParity(stereo: number): number | null {
  let newAtomParity: number | null = null;
  switch (stereo) {
    case Bond.PATTERN.STEREO.UP:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.ODD;
      break;
    case Bond.PATTERN.STEREO.EITHER:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EITHER;
      break;
    case Bond.PATTERN.STEREO.DOWN:
      newAtomParity = Atom.PATTERN.STEREO_PARITY.EVEN;
      break;
  }
  return newAtomParity;
}
