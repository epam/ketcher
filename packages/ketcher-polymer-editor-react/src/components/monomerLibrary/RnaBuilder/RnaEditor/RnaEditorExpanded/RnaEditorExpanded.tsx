/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import { MonomerGroups } from 'src/constants';
import { GroupBlock } from './GroupBlock';
import {
  ButtonsContainer,
  GroupsContainer,
  NameContainer,
  NameInput,
  NameLine,
  RnaEditorExpandedContainer,
  StyledButton,
} from './styles';
import { IRnaEditorExpandedProps } from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditorExpanded/types';
import { useState } from 'react';

export const RnaEditorExpanded = ({
  name,
  onCancel,
  onChangeName,
}: IRnaEditorExpandedProps) => {
  const groupsData = [
    {
      groupName: MonomerGroups.SUGARS,
      iconName: 'sugar',
    },
    {
      groupName: MonomerGroups.BASES,
      iconName: 'base',
    },
    {
      groupName: MonomerGroups.PHOSPHATES,
      iconName: 'phosphate',
    },
  ] as const;

  const [selectedGroup, setSelectedGroup] = useState<MonomerGroups>();

  const selectGroup = (selectedGroup) => {
    setSelectedGroup(selectedGroup);
  };

  return (
    <RnaEditorExpandedContainer>
      <NameContainer>
        <NameInput
          value={name}
          placeholder="Name your structure"
          onChange={onChangeName}
        />
        <NameLine selected />
      </NameContainer>
      <GroupsContainer>
        {groupsData.map(({ groupName, iconName }) => {
          return (
            <GroupBlock
              key={groupName}
              selected={selectedGroup === groupName}
              groupName={groupName}
              iconName={iconName}
              onClick={() => selectGroup(groupName)}
            />
          );
        })}
      </GroupsContainer>
      <ButtonsContainer>
        <StyledButton onClick={onCancel}>Cancel</StyledButton>
        <StyledButton disabled>Add to Presets</StyledButton>
      </ButtonsContainer>
    </RnaEditorExpandedContainer>
  );
};
