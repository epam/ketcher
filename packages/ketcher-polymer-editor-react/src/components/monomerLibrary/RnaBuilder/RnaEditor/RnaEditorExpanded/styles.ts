/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import styled from '@emotion/styled';
import { Button, Input } from 'ketcher-react';

export const RnaEditorExpandedContainer = styled.div({
  display: 'flex',
  flexDirection: 'column',
  padding: '10px',
});

export const GroupsContainer = styled.div({
  marginTop: '16px',
  width: '100%',
  display: 'flex',
  flexDirection: 'column',
  gap: '8px',
});

export const ButtonsContainer = styled.div({
  marginTop: '16px',
  display: 'flex',
  justifyContent: 'space-evenly',
  alignItems: 'center',
  gap: '8px',
});

export const StyledButton = styled(Button)((props) => ({
  width: '100%',
  whiteSpace: 'nowrap',
  fontSize: props.theme.ketcher.font.size.regular,
}));

export const NameContainer = styled.div<{ selected?: boolean }>((props) => ({
  position: 'relative',
  borderRadius: props.theme.ketcher.border.radius.regular,
  backgroundColor: props.theme.ketcher.color.background.primary,
  boxShadow: props.theme.ketcher.shadow.regular,
  overflow: 'hidden',
  padding: '28px 6px 6px 6px',
  outline: props.selected
    ? props.theme.ketcher.outline.selected.medium
    : 'none',

  ':hover': {
    outline: props.selected ? undefined : props.theme.ketcher.outline.small,
  },
}));

export const NameLine = styled.span<{ selected?: boolean }>((props) => ({
  position: 'absolute',
  top: '0',
  left: '0',
  width: '100%',
  height: '8px',
  backgroundColor: props.selected
    ? props.theme.ketcher.outline.selected.color
    : props.theme.ketcher.outline.color,
}));

export const NameInput = styled(Input)({
  width: '100%',
});
