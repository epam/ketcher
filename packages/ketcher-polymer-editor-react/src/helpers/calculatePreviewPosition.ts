/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { EditorClassName, preview } from '../constants';
import { MonomerItemType } from 'ketcher-core';

export const calculatePreviewPosition = (
  monomer: MonomerItemType | undefined,
  target?: DOMRect,
): string => {
  if (monomer && target) {
    const editorRect = document
      .querySelector(`.${EditorClassName}`)
      ?.getBoundingClientRect();

    if (!editorRect || !target) {
      return '';
    }

    const top =
      target.top > preview.height + preview.gap
        ? target.top - preview.gap - preview.height - preview.topPadding
        : target.bottom + preview.gap - preview.topPadding;

    const newStyle = `${top}px`;

    return newStyle;
  }
  return '';
};
