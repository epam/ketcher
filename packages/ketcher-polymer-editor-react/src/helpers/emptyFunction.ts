// eslint-disable-next-line @typescript-eslint/no-empty-function
export const EmptyFunction = () => {};
