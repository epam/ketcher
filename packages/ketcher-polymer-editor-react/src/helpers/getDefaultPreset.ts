import { IRnaPreset } from 'components/monomerLibrary/RnaBuilder/types';
import { MonomerItemType } from 'ketcher-core';

const defaultPresetBases = {
  A: 'Adenine',
  C: 'Cytosine',
  G: 'Guanine',
  T: 'Thymine',
  U: 'Uracil',
};
const SUGAR = 'Ribose';
const PHOSPHATE = 'Phosphate';

export const getDefaultPresets = (
  monomers: MonomerItemType[],
): IRnaPreset[] => {
  const defaultBaseNames = Object.values(defaultPresetBases);
  const ribose = monomers.find((item) => item.props.Name === SUGAR);
  const phosphate = monomers.find((item) => item.props.Name === PHOSPHATE);
  const defaultBases = monomers.filter((item) => {
    const base = defaultBaseNames.find(
      (defaultName) => defaultName === item.props.Name,
    );
    return !!base;
  });
  let presets;
  if (ribose && phosphate) {
    presets = defaultBases.map((base) => {
      const nucleotide: IRnaPreset = {
        base,
        sugar: ribose,
        phosphate,
        name: base.props.MonomerName,
      };
      return nucleotide;
    });
  } else {
    presets = [];
  }
  return presets;
};
