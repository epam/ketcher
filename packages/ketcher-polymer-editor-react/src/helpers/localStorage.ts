class LocalStorageWrapper {
  localStorage: Storage;

  constructor() {
    this.localStorage = window.localStorage;
  }

  getItem(key: string) {
    const item: string | null = this.localStorage.getItem(key);

    if (!item) {
      return null;
    }

    return JSON.parse(item);
  }

  setItem(key: string, item: unknown) {
    this.localStorage.setItem(key, JSON.stringify(item));
  }
}

export const localStorageWrapper = new LocalStorageWrapper();
