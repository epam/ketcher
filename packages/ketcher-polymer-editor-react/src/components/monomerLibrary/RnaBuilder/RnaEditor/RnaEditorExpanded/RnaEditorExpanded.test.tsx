import { fireEvent, render, screen } from '@testing-library/react';
import { RnaEditorExpanded } from 'components/monomerLibrary/RnaBuilder/RnaEditor/RnaEditorExpanded/RnaEditorExpanded';
import { EmptyFunction } from 'helpers';

describe('Test Rna Editor Expanded component', () => {
  it('should render correctly in edit mode', async () => {
    const onCancelHandler = jest.fn();
    const onSaveHandler = jest.fn();

    render(
      withThemeAndStoreProvider(
        <RnaEditorExpanded
          isEditMode
          onCancel={onCancelHandler}
          onChangeName={EmptyFunction}
          onDuplicate={EmptyFunction}
          onEdit={EmptyFunction}
          onSave={onSaveHandler}
        />,
        {
          rnaBuilder: {
            activePreset: {
              name: 'MyRna',
              sugar: {
                props: {
                  MonomerName: '',
                },
              },
              phosphate: {
                props: {
                  MonomerName: '',
                },
              },
              base: {
                props: {
                  MonomerName: '',
                },
              },
            },
          },
        },
      ),
    );

    const rnaEditorExpanded = screen.getByTestId('rna-editor-expanded');
    const cancelBtn = screen.getByTestId('cancel-btn');
    const addToPresetsBtn = screen.getByTestId('add-to-presets-btn');

    fireEvent.click(addToPresetsBtn);
    fireEvent.click(cancelBtn);

    expect(onSaveHandler).toBeCalled();
    expect(onCancelHandler).toBeCalled();
    expect(rnaEditorExpanded).toMatchSnapshot();
  });

  it('should render correctly in view mode', async () => {
    const onDuplicateHandler = jest.fn();
    const onEditHandler = jest.fn();

    render(
      withThemeAndStoreProvider(
        <RnaEditorExpanded
          isEditMode={false}
          onCancel={EmptyFunction}
          onChangeName={EmptyFunction}
          onDuplicate={onDuplicateHandler}
          onEdit={onEditHandler}
          onSave={EmptyFunction}
        />,
        {
          rnaBuilder: {
            activePreset: {
              name: 'MyRna',
              sugar: {
                props: {
                  MonomerName: '',
                },
              },
              phosphate: {
                props: {
                  MonomerName: '',
                },
              },
              base: {
                props: {
                  MonomerName: '',
                },
              },
              presetInList: {},
            },
          },
        },
      ),
    );

    const rnaEditorExpanded = screen.getByTestId('rna-editor-expanded');
    const editBtn = screen.getByTestId('edit-btn');
    const duplicateBtn = screen.getByTestId('duplicate-btn');

    fireEvent.click(editBtn);
    fireEvent.click(duplicateBtn);

    expect(onEditHandler).toBeCalled();
    expect(onDuplicateHandler).toBeCalled();
    expect(rnaEditorExpanded).toMatchSnapshot();
  });
});
