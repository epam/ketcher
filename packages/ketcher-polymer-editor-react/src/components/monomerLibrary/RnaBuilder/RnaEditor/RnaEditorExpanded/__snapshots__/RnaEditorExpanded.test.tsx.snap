// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Rna Editor Expanded component should render correctly in edit mode 1`] = `
<div
  class="css-zbq0po"
  data-testid="rna-editor-expanded"
>
  <div
    class="css-19lk7et"
  >
    <input
      class="css-9fgya1"
      placeholder="Name your structure"
      value=""
    />
    <span
      class="css-1o94icx"
    />
  </div>
  <div
    class="css-s3o4ys"
  >
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="13.333"
          rx="1"
          width="13.333"
          x="1.332"
          y="1.333"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Sugar
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="17"
        viewBox="0 0 18 17"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="12"
          rx="1"
          transform="rotate(45 9 0)"
          width="12"
          x="9"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Base
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="20"
          rx="10"
          width="20"
          x="2"
          y="2"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Phosphate
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
  </div>
  <div
    class="css-1iug9fi"
  >
    <button
      class="css-1971dyf"
      data-testid="cancel-btn"
    >
      Cancel
    </button>
    <button
      class="css-4dyvry"
      data-testid="add-to-presets-btn"
    >
      Add to Presets
    </button>
  </div>
</div>
`;

exports[`Test Rna Editor Expanded component should render correctly in view mode 1`] = `
<div
  class="css-zbq0po"
  data-testid="rna-editor-expanded"
>
  <div
    class="css-19lk7et"
  >
    <div
      class="css-1o4wo1x"
    />
    <span
      class="css-1o94icx"
    />
  </div>
  <div
    class="css-s3o4ys"
  >
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="13.333"
          rx="1"
          width="13.333"
          x="1.332"
          y="1.333"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Sugar
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="17"
        viewBox="0 0 18 17"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="12"
          rx="1"
          transform="rotate(45 9 0)"
          width="12"
          x="9"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Base
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
    <div
      class="css-1b10s7n"
    >
      <svg
        class="css-dsqlr9"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="currentColor"
          height="20"
          rx="10"
          width="20"
          x="2"
          y="2"
        />
      </svg>
      <div
        class="css-j7qwjs"
      >
        <span
          class="css-9c7r58"
        >
          Phosphate
        </span>
        <span
          class="css-1mtxt16"
        >
          Not selected
        </span>
      </div>
    </div>
  </div>
  <div
    class="css-1iug9fi"
  >
    <button
      class="css-1971dyf"
      data-testid="duplicate-btn"
    >
      Duplicate and Edit
    </button>
    <button
      class="css-1971dyf"
      data-testid="edit-btn"
    >
      Edit
    </button>
  </div>
</div>
`;
